FileIO/MPI_IO/NodePartitionedMeshReader.cpp

```diff
@@ -0,0 +1,521 @@
+
+    if(!file_opened)
+        return nullptr;
+
+    MeshLib::NodePartitionedMesh* np_mesh = nullptr;
+    std::vector<MeshLib::Node*> mesh_nodes;
+    std::vector<unsigned long> glb_node_ids;
+    std::vector<MeshLib::Element*> mesh_elems;
+
+    for(int i = 0; i < _mpi_comm_size; i++)
+    {
+        if(_mpi_rank == 0)
+        {
+            INFO("-->Parallel reading the partitioned mesh: ");
+
+            // Read first part into _mesh_info which is equal with the binary
+            // structure.
+            for(unsigned long j = 0; j < 10; ++j)
+                is_cfg >> *(_mesh_info.data() + j);
+            // The last positon is the extra_flag.
+            is_cfg >> _mesh_info.extra_flag;
+            is_cfg >> std::ws;
+        }
+
+        MPI_Bcast(_mesh_info.data(), static_cast<int>(_mesh_info.size()),
+            MPI_LONG, 0, _mpi_comm);
+
+        //----------------------------------------------------------------------------------
+        // Read Nodes
+        if (!readCastNodesASCII(is_node, i, mesh_nodes, glb_node_ids))
+            break;
+
+        //----------------------------------------------------------------------------------
+        // Read elements
+        if (!readCastElemsASCII(is_elem, i,
+            _mesh_info.regular_elements + _mesh_info.offset[0],
+            false, mesh_nodes, mesh_elems))
+            break;
+
+        //-------------------------------------------------------------------------
+        // Ghost elements
+        if (!readCastElemsASCII(is_elem, i,
+            _mesh_info.ghost_elements + _mesh_info.offset[1],
+            true, mesh_nodes, mesh_elems))
+            break;
+
+        if(_mpi_rank == i)
+            np_mesh = newMesh(BaseLib::extractBaseName(file_name_base),
+                    mesh_nodes, glb_node_ids, mesh_elems);
+    }
+
+    if(_mpi_rank == 0)
+    {
+        is_cfg.close();
+        is_node.close();
+        is_elem.close();
+    }
+
+    MPI_Barrier(_mpi_comm);
+
+    return np_mesh;
+}
+
+MeshLib::NodePartitionedMesh*
+NodePartitionedMeshReader::newMesh(
+    std::string const& mesh_name,
+    std::vector<MeshLib::Node*> const& mesh_nodes,
+    std::vector<unsigned long> const& glb_node_ids,
+    std::vector<MeshLib::Element*> const& mesh_elems) const
+{
+    return new MeshLib::NodePartitionedMesh(
+        mesh_name + std::to_string(_mpi_comm_size),
+        mesh_nodes, glb_node_ids, mesh_elems,
+        _mesh_info.regular_elements,
+        _mesh_info.global_base_nodes,
+        _mesh_info.global_nodes,
+        _mesh_info.base_nodes,
+        _mesh_info.active_base_nodes,
+        _mesh_info.active_nodes);
+}
+
+void NodePartitionedMeshReader::readElementASCII(std::ifstream &ins,
+    std::vector<unsigned long>& elem_data, const bool ghost) const
+{
+    // Set number of elements.
+    unsigned long const ne =
+        ghost ? _mesh_info.ghost_elements : _mesh_info.regular_elements;
+    unsigned long counter = ne;
```

Comment:
From entry ID ``ne``, the number of element, array ``elem_data`` saves the element data.  For any j in  ``[0: ne-1]``, array ``elem_data[j]` saves the starting positions of the data for each element (also see my reply below).